File: Data Structures & Algorithms/merge-intervals/submission-1.py
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        intervals.sort(key=lambda pair: pair[0])
        res = [intervals[0]]

        for a, b in intervals:
            lastEnd = res[-1][1]

            if lastEnd >= a:
                res[-1][1] = max(b, lastEnd)
            else:
                res.append([a, b])
        return res

File: Data Structures & Algorithms/insert-new-interval/submission-1.cs
public class Solution {
    public int[][] Insert(int[][] intervals, int[] newInterval) {
        List<int[]> res = new List<int[]>();

        int i = 0;
        int n = intervals.Count();
        /*
            ....       .....       ......  [a, b]  a <= d && c <= b
                  ....    ....       ...   [c, d]
        */ 
        while (i < n && intervals[i][1] < newInterval[0]) {
            res.Add(intervals[i]);
            i += 1;
        }
        // b only gets bigger every time given that the intervals are sorted
        // and c becomes the same or smaller in every iteration of the loop.
        while (i < n && intervals[i][0] <= newInterval[1]) {
            newInterval[0] = Math.Min(intervals[i][0], newInterval[0]);
            newInterval[1] = Math.Max(intervals[i][1], newInterval[1]);
            i += 1;
        }
        res.Add(newInterval);

        while (i < n) {
            res.Add(intervals[i]);
            i += 1;
        }

        return res.ToArray();
    }
}
